editor/create_dialog.cpp

```diff
@@ -169,7 +171,94 @@ bool CreateDialog::_should_hide_type(const String &p_type) const {
 	return false;
 }
 
-void CreateDialog::_update_search() {
+void CreateDialog::_search_thread_func(void *p_userdata) {
+	CreateDialog *dialog = (CreateDialog *)p_userdata;
+	while (true) {
+		dialog->search_invalidate.clear();
+		bool search_invalidated = false;
+
+		const String search_text = dialog->search_box->get_text();
+		bool empty_search = search_text.is_empty();
+
+		// Filter all candidate results.
+		Vector<String> candidates;
+		for (List<StringName>::Element *I = dialog->type_list.front(); I; I = I->next()) {
+			if (dialog->search_abort.is_set())
+				return;
+			if (dialog->search_invalidate.is_set()) {
+				search_invalidated = true;
+				break;
+			}
+			if (empty_search || search_text.is_subsequence_ofn(I->get())) {
+				candidates.push_back(I->get());
+			}
+		}
+		if (search_invalidated) {
+			continue;
+		}
+
+		// Build the type tree.
+		for (int i = 0; i < candidates.size(); i++) {
+			if (dialog->search_abort.is_set())
+				return;
+			if (dialog->search_invalidate.is_set()) {
+				search_invalidated = true;
+				break;
+			}
+			dialog->_add_type(candidates[i], ClassDB::class_exists(candidates[i]) ? TypeCategory::CPP_TYPE : TypeCategory::OTHER_TYPE);
+		}
+		if (search_invalidated) {
+			continue;
+		}
+
+		// Configure search items.
+		callable_mp(dialog, &CreateDialog::_create_search_option_items).bind(candidates, search_text).call_deferred();
+
+		while (true) {
```

Comment:
This sort of busy loop is detrimental to performance. I'd suggest you to use a semaphore or a condition variable instead.